data/src/main/java/com/microsoft/azure/kusto/data/AadAuthenticationHelper.java

```diff
@@ -3,300 +3,282 @@
+                return lastConfidentialClientApplication.acquireTokenSilently(getSilentParameters(accounts)).get(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+            case PUBLIC:
+                accounts = lastPublicClientApplication.getAccounts();
+                return lastPublicClientApplication.acquireTokenSilently(getSilentParameters(accounts)).get(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+            case NOT_APPLICABLE:
+                throw new DataClientException("Cannot obtain a refresh token for Authentication type '" + authenticationType.name());
+            default:
+                throw new DataClientException("Authentication type '" + authenticationType.name() + "' is invalid");
         }
     }
 
-    private boolean IsInvalidToken() {
-        return lastAuthenticationResult == null || lastAuthenticationResult.getExpiresOnDate().before(dateInAMinute());
-    }
+    private SilentParameters getSilentParameters(CompletableFuture<Set<IAccount>> accounts) {
+        IAccount account;
+        if (StringUtils.isNotBlank(userUsername)) {
+            account = accounts.join().stream().filter(u -> userUsername.equalsIgnoreCase(u.username())).findAny().orElse(null);
+        } else {
+            // Normally we would filter accounts by the user authenticating, but there's only 1 per AadAuthenticationHelper instance
+            account = accounts.join().iterator().next();
+        }
 
-    AuthenticationResult acquireAccessTokenByRefreshToken() throws DataServiceException {
-        AuthenticationContext context;
-        ExecutorService service = null;
+        return SilentParameters.builder(scopes).account(account).authorityUrl(aadAuthorityUrl).build();
+    }
 
-        try {
-            service = Executors.newSingleThreadExecutor();
-            context = new AuthenticationContext(aadAuthorityUri, false, service);
-            switch (authenticationType) {
-                case AAD_APPLICATION_KEY:
-                case AAD_APPLICATION_CERTIFICATE:
-                    return context.acquireTokenByRefreshToken(lastAuthenticationResult.getRefreshToken(), clientCredential, null).get();
-                case AAD_USERNAME_PASSWORD:
-                case AAD_DEVICE_LOGIN:
-                    return context.acquireTokenByRefreshToken(lastAuthenticationResult.getRefreshToken(), CLIENT_ID, clusterUrl, null).get();
-                default:
-                    throw new DataServiceException("Authentication type: " + authenticationType.name() + " is invalid");
-            }
-        } catch (Exception e) {
-            throw new DataServiceException(e.getMessage(), e);
-        } finally {
-            if (service != null) {
-                service.shutdown();
-            }
+    // TODO: I see a valuable performance improvement by freeing this thread from blocking to do other things (perhaps the user is concurrently ingesting?). I suspect others will consider this overkill.
```

Comment:
Reviewers: Point for discussion

Reply:
I think there are 2 relevant concepts here:
1) Blocking - synchronous code that prevents the *next* line of code from running. Indeed, we cannot return the execution context to the SDK-user until we authenticate (they're likely to make a privileged call next). I agree on this.
2) Preventing (I shouldn't have used the term "blocking") another execution context from utilizing the CPU resources that this thread is using (see https://docs.oracle.com/javase/tutorial/essential/concurrency/sleep.html).

Here, I'm trying to improve performance via (2). Admittedly, it might not be very much/worth it in practice.

Reply:
It's uncommon to address (2), so I removed this since the performance benefit is hard to quantify because it's entirely dependent on the provided token provider, and in the interest of keeping the code cleaner.